contracts/MarginCalculator.sol

```diff
@@ -302,53 +332,56 @@ contract MarginCalculator {
         );
 
         // ensure the long asset is valid for the short asset
-        require(_isMarginableLong(_vault), "MarginCalculator: long asset not marginable for short asset");
+        require(
+            _isMarginableLong(_vault, _vaultDetails),
+            "MarginCalculator: long asset not marginable for short asset"
+        );
 
         // ensure that the collateral asset is valid for the short asset
-        require(_isMarginableCollateral(_vault), "MarginCalculator: collateral asset not marginable for short asset");
+        require(
+            _isMarginableCollateral(_vault, _vaultDetails),
+            "MarginCalculator: collateral asset not marginable for short asset"
+        );
     }
 
     /**
      * @dev if there is a short option and a long option in the vault, ensure that the long option is able to be used as collateral for the short option
      * @param _vault the vault to check.
      */
-    function _isMarginableLong(MarginVault.Vault memory _vault) internal view returns (bool) {
-        bool hasLong = _isNotEmpty(_vault.longOtokens);
-        bool hasShort = _isNotEmpty(_vault.shortOtokens);
+    function _isMarginableLong(MarginVault.Vault memory _vault, VaultDetails memory _vaultDetails)
+        internal
+        view
+        returns (bool)
+    {
         // if vault is missing a long or a short, return True
-        if (!hasLong || !hasShort) return true;
-
-        OtokenInterface long = OtokenInterface(_vault.longOtokens[0]);
-        OtokenInterface short = OtokenInterface(_vault.shortOtokens[0]);
+        if (!_vaultDetails.hasLong || !_vaultDetails.hasShort) return true;
 
         return
             _vault.longOtokens[0] != _vault.shortOtokens[0] &&
-            long.underlyingAsset() == short.underlyingAsset() &&
-            long.strikeAsset() == short.strikeAsset() &&
-            long.collateralAsset() == short.collateralAsset() &&
-            long.expiryTimestamp() == short.expiryTimestamp() &&
-            long.isPut() == short.isPut();
+            _vaultDetails.longUnderlyingAsset == _vaultDetails.shortUnderlyingAsset &&
+            _vaultDetails.longStrikeAsset == _vaultDetails.shortStrikeAsset &&
+            _vaultDetails.longCollateralAsset == _vaultDetails.shortCollateralAsset &&
+            _vaultDetails.longExpiryTimestamp == _vaultDetails.shortExpiryTimestamp &&
+            _vaultDetails.longType == _vaultDetails.shortType;
     }
```

Comment:
didn't realize how expensive this was until looking at this diff 😮

Reply:
yeah external calls are 💸